ql/src/test/results/clientpositive/llap/subquery_ALL.q.out

```diff
@@ -413,8 +413,7 @@ POSTHOOK: Input: default@part
 POSTHOOK: Input: default@part_null_n0
 #### A masked pattern was here ####
 26
-Warning: Shuffle Join MERGEJOIN[37][tables = [$hdt$_1, $hdt$_2]] in Stage 'Reducer 3' is a cross product
-Warning: Shuffle Join MERGEJOIN[38][tables = [$hdt$_1, $hdt$_2, $hdt$_0]] in Stage 'Reducer 4' is a cross product
+Warning: Shuffle Join MERGEJOIN[22][tables = [$hdt$_0, $hdt$_1]] in Stage 'Reducer 2' is a cross product
```

Comment:
Will these changes work for insert statements that produce zero records also, or only for select queries?
I am asking because there is a ticket https://issues.apache.org/jira/browse/HIVE-24466 - "insert queries should not launch job when condition in the query would output 0 rows", I was wondering it these changes are a fix for that ticket too.

Reply:
> Is there a branch here that is simplified to empty? Why?

This is the plan after subquery rewrite and decorrelation:
```
HiveProject(_o__c0=[$0])
  HiveAggregate(group=[{}], agg#0=[count()])
    HiveProject($f0=[$0])
      HiveProject(p_partkey=[$0], p_name=[$1], p_mfgr=[$2], p_brand=[$3], p_type=[$4], p_size=[$5], p_container=[$6], p_retailprice=[$7], p_comment=[$8], BLOCK__OFFSET__INSIDE__FILE=[$9], INPUT__FILE__NAME=[$10], ROW__ID=[$11], ROW__IS__DELETED=[$12])
        HiveFilter(condition=[IS NULL(OR(AND(IS NOT NULL($16), <>($13, 0)), AND(OR(IS NULL($0), <($14, $13)), null, <>($13, 0), IS NULL($16))))])
          HiveJoin(condition=[=($0, $15)], joinType=[left], algorithm=[none], cost=[not available])
            HiveJoin(condition=[true], joinType=[inner], algorithm=[none], cost=[not available])
              HiveTableScan(table=[[default, part]], table:alias=[part])
              HiveAggregate(group=[{}], c=[COUNT()], ck=[COUNT($0)])
                HiveProject(p_partkey=[$0])
                  HiveFilter(condition=[IS NULL($0)])
                    HiveTableScan(table=[[default, part_null_n0]], table:alias=[part_null_n0])
            HiveAggregate(group=[{0, 1}])
              HiveProject(p_partkey=[$0], literalTrue=[true])
                HiveFilter(condition=[IS NULL($0)])
                  HiveTableScan(table=[[default, part_null_n0]], table:alias=[part_null_n0])
```
`HiveJoin(condition=[=($0, $15)], joinType=[left]` is removed because ReduceExpressionRule transforms the join condition always `false`.
It does because `$15` coming from the right branch is always `null` due to `HiveFilter(condition=[IS NULL($0)])`

Reply:
@difin 
The plan generated for the select part is simplified but it will be executed. Even if the outer query limit is set to 0
https://github.com/apache/hive/blob/70562437d369c2f4ab3e879bae519f81d386da3b/ql/src/java/org/apache/hadoop/hive/ql/parse/TaskCompiler.java#L138

The optimization is enabled only for queries
https://github.com/apache/hive/blob/70562437d369c2f4ab3e879bae519f81d386da3b/ql/src/java/org/apache/hadoop/hive/ql/parse/TaskCompiler.java#L182